src/client/cmd/common_cli.cpp

```diff
@@ -0,0 +1,122 @@
+ *
+ */
+
+#include "common_cli.h"
+#include "exec.h"
+
+#include <multipass/cli/argparser.h>
+#include <multipass/cli/format_utils.h>
+
+#include <fmt/ostream.h>
+
+namespace mp = multipass;
+namespace cmd = multipass::cmd;
+
+namespace
+{
+mp::ReturnCode return_code_for(const grpc::StatusCode& code)
+{
+    return code == grpc::StatusCode::UNAVAILABLE ? mp::ReturnCode::DaemonFail : mp::ReturnCode::CommandFail;
+}
+} // namespace
+
+mp::ParseCode cmd::handle_all_option(ArgParser* parser, std::ostream& cerr)
+{
+    auto num_names = parser->positionalArguments().count();
+    if (num_names == 0 && !parser->isSet(all_option_name))
+    {
+        fmt::print(cerr, "Name argument or --all is required\n");
+        return ParseCode::CommandLineError;
+    }
+
+    if (num_names > 0 && parser->isSet(all_option_name))
+    {
+        fmt::print(cerr, "Cannot specify name{} when --all option set\n", num_names > 1 ? "s" : "");
+        return ParseCode::CommandLineError;
+    }
+
+    return ParseCode::Ok;
+}
+
+mp::InstanceNames cmd::add_instance_names(ArgParser* parser)
+{
+    InstanceNames instance_names;
+
+    for (const auto& arg : parser->positionalArguments())
+    {
+        auto instance_name = instance_names.add_instance_name();
+        instance_name->append(arg.toStdString());
+    }
+
+    return instance_names;
+}
+
+mp::ParseCode cmd::handle_format_option(ArgParser* parser, Formatter** chosen_formatter, std::ostream& cerr)
+{
+    *chosen_formatter = mp::format::formatter_for(parser->value(format_option_name).toStdString());
+
+    if (*chosen_formatter == nullptr)
+    {
+        fmt::print(cerr, "Invalid format type given.\n");
+        return ParseCode::CommandLineError;
+    }
+
+    return ParseCode::Ok;
+}
+
+std::string cmd::instance_action_message_for(const InstanceNames& instance_names, const std::string& action_name)
+{
+    std::string message{action_name};
+
+    if (instance_names.instance_name().empty())
+        message.append("all instances");
+    else if (instance_names.instance_name().size() > 1)
+        message.append("requested instances");
+    else
+        message.append(instance_names.instance_name().Get(0));
+
+    return message;
+}
+
+mp::ReturnCode cmd::standard_failure_handler_for(const std::string& command, std::ostream& cerr,
+                                                 const grpc::Status& status, const std::string& error_details)
+{
+    fmt::print(cerr, "{} failed: {}\n{}", command, status.error_message(), error_details);
+
+    return return_code_for(status.error_code());
+}
+
+void cmd::install_sshfs_for(const std::string& instance_name, int verbosity_level, grpc::Channel* rpc_channel,
+                            Rpc::Stub* stub, std::ostream& cout, std::ostream& cerr)
+{
+    std::vector<Command::UPtr> command;
+    command.push_back(std::make_unique<Exec>(*rpc_channel, *stub, cout, cerr));
+
+    auto args = QStringList() << ""
```

Comment:
why have an empty string as first member of the QStringList?

Reply:
Since we are basically overloading the exec command here, that empty string represents the actual binary name in the list of arguments, like `multipass` when using it from the actual command line,  Since we don't care what it is here, I just used the empty string.  It could have been any string, like `foo` or `dummy` or something.  If it's clearer to put something here, I can do that.

Reply:
Ok, I suspected that. A quick comment explaining that it will being used like that would do the trick.